abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Creates metric (`metric_name_data`, `metic_name_size`) of type `metric_type`.
+
+It can be referred to in `proxy_get_metric_value`, `proxy_set_metric_value`,
+and `proxy_increment_metric_value` using returned unique metric identifier
+(`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `metric_type`.
+
+Returns `InvalidMemoryAccess` when `metric_name_data`, `metric_name_size`
+and/or `return_metric_id` point to invalid memory address.
+
+#### `proxy_get_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i32 (uint64_t*) return_value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get the value of the metric (`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+#### `proxy_set_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (uint64_t) value`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set the value of the metric (`metric_id`) to the value (`value`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when the `metric_type` doesn't support setting
+values.
+
+#### `proxy_increment_metric_value`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+  - `i64 (int64_t) offset`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Increment/decrement value of the metric (`metric_id`) by offset (`offset`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+Returns `InvalidOperation` when `metric_type` doesn't support incrementing
+and/or decrementing values.
+
+#### `proxy_delete_metric`
+
+* params:
+  - `i32 (uint32_t) metric_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created metric (`metric_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the metric (`metric_id`) doesn't exist.
+
+
+## HTTP callouts
+
+### Functions exposed by the host environment
+
+#### `proxy_dispatch_http_call`
+
+* params:
+  - `i32 (const char*) upstream_name_data`
+  - `i32 (size_t) upstream_name_size`
+  - `i32 (iovec_array_t[]) headers_map_data`
+  - `i32 (size_t) headers_map_size`
+  - `i32 (iovec_array_t) body_data`
+  - `i32 (size_t) body_size`
+  - `i32 (iovec_array_t[]) trailers_map_data`
+  - `i32 (size_t) trailers_map_size`
+  - `i32 (uint32_t) timeout_milliseconds`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Dispatches a HTTP call to upstream (`upstream_name_data`, `upstream_name_size`).
+
+Once the response is returned to the host, `proxy_on_http_call_response` will be
+called with an unique call identifier (`return_callout_id`).
+
+Returns `Ok` on success.
+
+Returns `NotFound` for unknown upstream.
+
+Returns `NotAllowed` when requests to given upstream are not allowed.
+
+Returns `InvalidMemoryAccess` when `upstream_name_data`, `upstream_name_size`,
```

Comment:
FWIW I learned that Envoy will return an error in this call if the ":path", ":method", and "host" headers aren't set. In general we should say something about what the various header fields mean to the host. Do we want to document that?